### ISSUE TITLE
Validation errors triggered by MsosStackModel

### ISSUE THREAD

**Issue by lznakano** [spacetelescope/roman_datamodels#229]
I installed romancal version 11, and added a function to convert the stack images to the MsosStackModel and then save as an asdf file.  Here is the piece of codes:

   ```
wcs = getWcs(**kwargs)
   meta['wcs']=wcs
   msosStack =  stnode.MsosStack()

   msosStack["meta"] = meta
   msosStack['data'] = fluxData  #u.Quantity(fluxData, ru.electron / ru.DN, dtype=np.float64)
   msosStack['uncertainty'] = unctData
   msosStack['mask'] = mask
   msosStack['coverage'] = coverageMap

   af = asdf.AsdfFile()
   msosStackModel = rdm.MsosStackModel(msosStack)
   af.tree = {'roman': msosStackModel}
```

line 294: msosStackModel = rdm.MsosStackModel(msosStack)

52b8672f1d4e:~/fits2Asdf# python3 fitsToasdf.py
Traceback (most recent call last):
  File "/home/fits2Asdf/fitsToasdf.py", line 428, in <module>
    fits2MsosStack(fluxfile, unctfile, outfile, **kwargs)
  File "/home/fits2Asdf/fitsToasdf.py", line 294, in fits2MsosStack
    msosStackModel = rdm.MsosStackModel(msosStack)
  File "/usr/lib/python3.10/site-packages/roman_datamodels/datamodels/_core.py", line 88, in __init__
    asdffile.tree = {"roman": init}
  File "/usr/lib/python3.10/site-packages/asdf/asdf.py", line 638, in tree
    self._validate(asdf_object, custom=bool(tree))
  File "/usr/lib/python3.10/site-packages/asdf/asdf.py", line 665, in _validate
    schema.validate(tagged_tree, self, reading=reading)
  File "/usr/lib/python3.10/site-packages/asdf/schema.py", line 677, in validate
    validator.validate(instance)
  File "/usr/lib/python3.10/site-packages/jsonschema/validators.py", line 314, in validate
    raise error
jsonschema.exceptions.ValidationError: 'calibration_software_version' is a required property

**Comment by WilliamJamieson** [comment]
You are not sufficiently constructing your `meta` object. Please use the `mk_msos_stack` utility.

**Comment by WilliamJamieson** [comment]
Alternately, you can use https://github.com/spacetelescope/roman_datamodels/blob/c4b5d46c2bf24b3d40923d77a5f50ab36d8ff461/src/roman_datamodels/maker_utils/__init__.py#L78-L94, via:
```python
from roman_datamodels.datatamodels import MsosStackModel
from roman_datamodels.maker_utils import mk_datamodel

kwargs = {
    "data": fluxData,
    "uncertainty": unctData,
    "mask": mask,
    "coverage": coverageMap
}

msosStackModel = mk_datamodel(MsosStackModel, **kwargs)
```

The `wcs` keyword is not part of the `meta` keywords, that information is stored under `wcsinfo` instead. @nden knows more about those specifics.

**Comment by WilliamJamieson** [comment]
Meta contains all of the metadata provided to all of the other datamodels plus a your image list file's path string.

Included within meta is all of the information you requested. You can find this information in the common-1.0.0 schema in RAD. The utility I linked you to will fill in all the information within meta with dummy information if you do not specifically override it with something different. If you don't want to override directly in the call to the utility you can instead simply set the fields in the resulting model to whatever you want. 

In general, I suggest you ignore any fields in that metadata which are not useful to you as the amount of extraneous information encoded in these extra fields is insignificant. This is what we do, as all of that information is provided and expected to be in any model (data file we will write and archive) we work with.

**Comment by lznakano** [comment]
Based on your comments, I don't need to create my own meta, until I need to set some of the values differently, right? That is great and makes my use easier.  Thanks!

**Comment by WilliamJamieson** [comment]
That is correct. The maker_utils can construct any of our datamodels or their component objects filled with dummy data (zero arrays, dummy strings, etc). This is so you can easily get a valid datamodel without having to tediously fill in everything. Theoretically, the final pipelines will include steps to properly populate all the fields when appropriate, but that is needlessly complicated for getting started.

I am putting together a full set of documentation about how roman_datamodels functions and all the useful utilities built into it.

Please feel free to ask more questions or request a teleconference with us if you need to. I will be happy to help you or find someone who can.

**Comment by lznakano** [comment]
Great work!!!   I had my script running nicely.  It is much simpler than what I did before.  I can update any meta information easily.  It will be really helpful if you can provide with some documents.  

Since wcsinfo will provides with enough information, I wonder if we really need to add wcs to the meta.   I now got some issue with adding the wcs.  

Thank you very much for your kind help.

**Comment by WilliamJamieson** [comment]
@lznakano, we are working to add a `wcs` keyword which would point directly at the GWCS object describing the WCS. I will add that keyword to your schema as part of those changes. For the time being, feel free to use:
```python
msosStackModel.meta["wcs"] = wcs
```
to include your WCS object in the metadata. As part of those changes I will work with you to get a WCS object which we can properly validate.

**Comment by lznakano** [comment]
> @lznakano, we are working to add a `wcs` keyword which would point directly at the GWCS object describing the WCS. I will add that keyword to your schema as part of those changes. For the time being, feel free to use:
> 
> ```python
> msosStackModel.meta["wcs"] = wcs
> ```
> 
> to include your WCS object in the metadata. As part of those changes I will work with you to get a WCS object which we can properly validate.

Thank you very much.

**Comment by lznakano** [comment]
@WilliamJamieson  Is the wcs added to the meta?  Can you please update me on the status?  Thanks!

**Comment by WilliamJamieson** [comment]
> @WilliamJamieson  Is the wcs added to the meta?  Can you please update me on the status?  Thanks!

Currently, there are upstream changes we are waiting on in order to get the wcs changes in here.

**Comment by lznakano** [comment]
@nden 
OK, I will give it a try then and let you know the result.

**Comment by lznakano** [comment]
I just tested it, here is the error I got:



msosStackModel.meta.wcs = wcs
  File "/Users/zhang/miniconda3/lib/python3.9/site-packages/roman_datamodels/stnode/_node.py", line 156, in __setattr__
    raise AttributeError(f"No such attribute ({key}) found in node")
AttributeError: No such attribute (wcs) found in node


Thanks!

Lijun

On Jul 24, 2023, at 8:29 AM, Nadia Dencheva ***@***.******@***.***>> wrote:



@lznakano<https://github.com/lznakano> I believe you should be able to do msosStackModel.meta.wcs = wcs.
If you are saying this does not work, can you provide a sample file or example code?

**Comment by nden** [comment]
@lznakano You didn't post your command. Are you using `msos.meta.wcs=wcs`? If so please try `msos.meta['wcs']=wcs`. We need to fix the former but the latter works for me.

**Comment by lznakano** [comment]
I just copied the line from the comments in the ticket and tested it.   I just tried what you listed here, and yes, it did work.

Thanks!

Lijun

On Aug 3, 2023, at 11:56 AM, Nadia Dencheva ***@***.******@***.***>> wrote:



@lznakano<https://github.com/lznakano> You didn't post your command. Are you using msos.meta.wcs=wcs? If so please try msos.meta['wcs']=wcs. We need to fix the former but the latter works for me.

**Comment by lznakano** [comment]
Although the wcs was added without error message,  it did not show in the meta.

 # this is the function to create the MsosStatck object and then add the wcs to the meta
 
 fits2MsosStack(fluxfile, unctfile, outfile, **kwargs)
 
 # open it using datamodels
  dm=rdm.open(outfile)
  dm.info(max_rows=150)
  
  meta = dm.meta
  keys=meta.keys()
  print(keys) 
  wcs = meta['wcs']

File "/Users/zhang/roman-msos-photometry/asdf_io/fits2Asdf/fitsToMsosStack.py", line 171, in <module>
    wcs = meta['wcs']
  File "/Users/zhang/miniconda3/lib/python3.9/site-packages/roman_datamodels/stnode/_node.py", line 209, in __getitem__
    raise KeyError(f"No such key ({key}) found in node")
KeyError: 'No such key (wcs) found in node'

**Comment by nden** [comment]
You are right. I can reproduce this.

**Comment by WilliamJamieson** [comment]
Try using the node object itself. So `mdl._instance.meta["wcs"] = ...`

If `wcs` was defined and initialized by the schema then what you were doing would work.

Once we properly add the `wcs` to all the schemas what you were trying to do should work.

**Comment by lznakano** [comment]
It worked now:

def fits2MsosStack(...):
   ....
   msosStackModel = mk_datamodel(MsosStackModel, **args)
   #msosStackModel._instance.meta["wcs"] = wcs

   wcs = getWcs(**kwargs)
   msosStackModel.meta['wcs'] = wcs
   msosStackModel.to_asdf(outfile)


if __name__ == "__main__":
   ....
   ....
   fits2MsosStack(fluxfile, unctfile, outfile, **kwargs)

   dm=rdm.open(outfile)
   dm.info(max_rows=150)

    meta = dm.meta
    keys=list(meta.keys())
    print(keys)
   wcs = meta['wcs']
   print(wcs.to_fits_sip())


I used  romandatamodels 0.17.0

Thank you very much for your support.
